### PR TITLE
Use ubuntu-drivers to install Nvidia drivers

### LINKED ISSUE #2042: Nvidia utils
For work using Nvidia GPU instances workspaces will need the Nvidia drivers and other packages provided by `nvidia-utils`.

There are multiple `nvidia-utils` packages for different driver/CUDA/... versions.
We should aim to install the most appropriate version for the hardware being used.
That might be a little tricky as it would depend on the VM SKU.

We could potentially use the VM size argument to choose which package version and template that in the desired state.
Or, we could have a map in the desired state and provide and argument (or variable or fact) to ansible to select the right version.

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [ ] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [ ] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [ ] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :closed_umbrella: Related issues

Closes #2042

### CODE REVIEW

**Review by jemrobinson — DISMISSED** [review]
We also need to install `ubuntu-drivers-common` for this to work. I don't think this is available for all the Ubuntu versions we're currently supporting.

**Comment by JimMadge** [comment]
The package is available for jammy and noble (at the moment we only _really_ support jammy).

File: data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml [hunk 1]
```diff
@@ -13,6 +13,7 @@ package_categories:
       - clamav
       - clamav-daemon
       - clamav-unofficial-sigs
+      - ubuntu-drivers-common
```
Comment (jemrobinson) [inline data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:16]: I don't think this is available for e.g. focal. Can you check?

Comment (JimMadge) [reply data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:16]: It might not be, but we don't support focal, only jammy.

**Comment by craddm** [comment]
```
Aug 06 11:03:23 shm-green-sre-moocow-vm-workspace-02 desired_state.sh[4946]: changed: [localhost] => (item={'name': 'pycharm-community', 'classic': True})
Aug 06 11:03:23 shm-green-sre-moocow-vm-workspace-02 desired_state.sh[4946]: TASK [Check for Nvidia drivers] ************************************************
Aug 06 11:03:24 shm-green-sre-moocow-vm-workspace-02 python3[37261]: ansible-ansible.builtin.apt Invoked with name=nvidia-utils state=present package=['nvidia-utils'] update_cache_retries=5 update_cache_retry_max_delay=12 cache_valid_time=0 purge=False force=False dpkg_options=force-confdef,force-confold autoremove=False autoclean=False only_upgrade=False force_apt_get=False allow_unauthenticated=False update_cache=None deb=None default_release=None install_recommends=None upgrade=None policy_rc_d=None
[ 1542.124465] desired_state.sh[4946]: fatal: [localhost]: FAILED! => {"cache_update_time": 1722941566, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"     --simulate install 'nvidia-utils'' failed: E: Package 'nvidia-utils' has no installation candidate\n", "rc": 100, "stderr": "E: Package 'nvidia-utils' has no installation candidate\n", "stderr_lines": ["E: Package 'nvidia-utils' has no installation candidate"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nPackage nvidia-utils is a virtual package provided by:\n  nvidia-utils-550-server 550.90.07-0ubuntu0.22.04.1\n  nvidia-utils-550 550.90.07-0ubuntu0.22.04.1\n  nvidia-utils-545 545.29.06-0ubuntu0.22.04.2\n  nvidia-utils-535-server 535.183.01-0ubuntu0.22.04.1\n  nvidia-utils-535 535.183.01-0ubuntu0.22.04.1\n  nvidia-utils-470-server 470.256.02-0ubuntu0.22.04.1\n  nvidia-utils-470 470.256.02-0ubuntu0.22.04.1\n  nvidia-utils-450-server 450.248.02-0ubuntu0.22.04.1\n  nvidia-utils-418-server 418.226.00-0ubuntu5~0.22.04.1\n  nvidia-utils-390 390.157-0ubuntu0.22.04.2\n\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "Package nvidia-utils is a virtual package provided by:", "  nvidia-utils-550-server 550.90.07-0ubuntu0.22.04.1", "  nvidia-utils-550 550.90.07-0ubuntu0.22.04.1", "  nvidia-utils-545 545.29.06-0ubuntu0.22.04.2", "  nvidia-utils-535-server 535.183.01-0ubuntu0.22.04.1", "  nvidia-utils-535 535.183.01-0ubuntu0.22.04.1", "  nvidia-utils-470-server 470.256.02-0ubuntu0.22.04.1", "  nvidia-utils-470 470.256.02-0ubuntu0.22.04.1", "  nvidia-utils-450-server 450.248.02-0ubuntu0.22.04.1", "  nvidia-utils-418-server 418.226.00-0ubuntu5~0.22.04.1", "  nvidia-utils-390 390.157-0ubuntu0.22.04.2", ""]}
[ 1542.132494] desired_state.sh[4946]: PLAY RECAP *********************************************************************
[ 1542.138688] desired_state.sh[4946]: localhost                  : ok=5    changed=2    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
[ 1542.382324] desired_state.sh[4945]: /
Aug 06 11:03:25 shm-green-sre-moocow-vm-workspace-02 desired_state.sh[4946]: fatal: [localhost]: FAILED! => {"cache_update_time": 1722941566, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"     --simulate install 'nvidia-utils'' failed: E: Package 'nvidia-utils' has no installation candidate\n", "rc": 100, "stderr": "E: Package 'nvidia-utils' has no installation candidate\n", "stderr_lines": ["E: Package 'nvidia-utils' has no installation candidate"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nPackage nvidia-utils is a virtual package provided by:\n  nvidia-utils-550-server 550.90.07-0ubuntu0.22.04.1\n  nvidia-utils-550 550.90.07-0ubuntu0.22.04.1\n  nvidia-utils-545 545.29.06-0ubuntu0.22.04.2\n  nvidia-utils-535-server 535.183.01-0ubuntu0.22.04.1\n  nvidia-utils-535 535.183.01-0ubuntu0.22.04.1\n  nvidia-utils-470-server 470.256.02-0ubuntu0.22.04.1\n  nvidia-utils-470 470.256.02-0ubuntu0.22.04.1\n  nvidia-utils-450-server 450.248.02-0ubuntu0.22.04.1\n  nvidia-utils-418-server 418.226.00-0ubuntu5~0.22.04.1\n  nvidia-utils-390 390.157-0ubuntu0.22.04.2\n\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "Package nvidia-utils is a virtual package provided by:", "  nvidia-utils-550-server 550.90.07-0ubuntu0.22.04.1", "  nvidia-utils-550 550.90.07-0ubuntu0.22.04.1", "  nvidia-utils-545 545.29.06-0ubuntu0.22.04.2", "  nvidia-utils-535-server 535.183.01-0ubuntu0.22.04.1", "  nvidia-utils-535 535.183.01-0ubuntu0.22.04.1", "  nvidia-utils-470-server 470.256.02-0ubuntu0.22.04.1", "  nvidia-utils-470 470.256.02-0ubuntu0.22.04.1", "  nvidia-utils-450-server 450.248.02-0ubuntu0.22.04.1", "  nvidia-utils-418-server 418.226.00-0ubuntu5~0.22.04.1", "  nvidia-utils-390 390.157-0ubuntu0.22.04.2", ""]}
```

**Comment by jemrobinson** [comment]
@craddm : I think ignoring this (expected) error makes the playbook run to completion.

**Comment by JimMadge** [comment]
That should work, feels a bit hacky. An alternative could be to use `ansible.builtin.stat`

**Comment by craddm** [comment]
testing now

**Comment by JimMadge** [comment]
Pushed a commit for the `stat` option, we can try both.

**Comment by JimMadge** [comment]
@craddm By the way, you don't need to redeploy to test. Just upload the next desired state stuff to the blob container.

**Comment by JimMadge** [comment]
Or... just use `creates:` 🤔

**Comment by JimMadge** [comment]
Working at [9bab3f0](https://github.com/alan-turing-institute/data-safe-haven/pull/2089/commits/9bab3f00bdabea4d9dc305c5ce0efdc03e6216d9)

With nvidia-utils

```
PLAY [Desired state configuration] *********************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [Use ubuntu-drivers to install Nvidia drivers] ****************************
ok: [localhost]
```

without

```
PLAY [Desired state configuration] *********************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [Use ubuntu-drivers to install Nvidia drivers] ****************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0  
```

**Review by jemrobinson — COMMENTED** [review]
LGTM but I haven't tested this. Have you @JimMadge or @craddm ?

**Comment by craddm** [comment]
I was just trying to test your fix, @jemrobinson, but if @JimMadge's works then...

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml [hunk 2]
```diff
@@ -30,6 +30,13 @@
         state: present
       loop: "{{ snap_packages }}"
 
+    # https://ubuntu.com/server/docs/nvidia-drivers-installation#installing-the-drivers-on-servers-andor-for-computing-purposes
+    - name: Use ubuntu-drivers to install Nvidia drivers  # noqa: no-handler
+      tags: nvidia
+      ansible.builtin.command:
+        cmd: ubuntu-drivers install --gpgpu
+        creates: /usr/bin/nvidia-smi
```
Comment (jemrobinson) [inline data_safe_haven/resources/workspace/ansible/desired_state.yaml:38]: Isn't this going to attempt an install on a non-GPU machine now? What happens then?

**Comment by jemrobinson** [comment]
@craddm : Mine did work (for me) but @JimMadge's made about 3 more versions since then :)

**Comment by craddm** [comment]
> @craddm By the way, you don't need to redeploy to test. Just upload the next desired state stuff to the blob container.

Once I've done that, how do I get the state to update?

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml (hunk 2 above)
Comment (JimMadge) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml:38]: Tested that, it reports "nothing to install"

**Comment by JimMadge** [comment]
@craddm You can run the desired-state service, the `/root/desired_state.sh` script or run ansible-playbook.

**Comment by craddm** [comment]
Ok, well James's fix worked for me too, but sounds like yours is the way to go

**Comment by JimMadge** [comment]
The current HEAD worked on my deployment ([see here](https://github.com/alan-turing-institute/data-safe-haven/pull/2089#issuecomment-2271245300)).

It isn't idempotent (or really, it will always report change when there is no driver to install), but that might be something we want to address later.

**Comment by JimMadge** [comment]
@craddm Do you want to test and/or leave a review?

**Review by craddm — APPROVED** [review]
Tested and works for me on a non-nvidia machine

**Pull request merged**